### PR TITLE
12.0 adyen statement import

### PR DESCRIPTION
**Pull request by MPishpecki**

### CODE REVIEW

File: account_bank_statement_import_adyen/tests/test_import_adyen.py [hunk 1]
```diff
@@ -51,69 +55,54 @@ def test01_import_adyen(self):
...
+            self._test_statement_import(
+                'account_bank_statement_import_adyen',
+                'adyen_test_invalid.xls',
+                'invalid', self.journal)
+
     def _test_statement_import(
-            self, module_name, file_name, statement_name, journal_id=False,
-            local_account=False, start_balance=False, end_balance=False,
-            transactions=None):
+            self, module_name, file_name, statement_name, journal_id=False):
```
Comment (MPishpecki) [inline account_bank_statement_import_adyen/tests/test_import_adyen.py]: journal_id is not needed any more

File: account_bank_statement_import_adyen/tests/test_import_adyen.py [hunk 2]
```diff
@@ -51,69 +55,54 @@ def test01_import_adyen(self):
...
             'adyen_test_credit_fees.xlsx',
             'YOURCOMPANY_ACCOUNT 2016/8', self.journal)
 
+    def test_03_import_adyen_invalid(self):
+        """ Trying to hit that coverall target """
+        with self.assertRaisesRegex(UserError, 'Could not make sense'):
+            self._test_statement_import(
+                'account_bank_statement_import_adyen',
+                'adyen_test_invalid.xls',
+                'invalid', self.journal)
```
Comment (MPishpecki) [inline account_bank_statement_import_adyen/tests/test_import_adyen.py]: self.journal not needed

File: account_bank_statement_import_adyen/tests/test_import_adyen.py [hunk 3]
```diff
@@ -51,69 +55,54 @@ def test01_import_adyen(self):
...
+        self.assertFalse(lines.mapped('matched_debit_ids'))
+        self.assertFalse(lines.mapped('matched_credit_ids'))
+        self.assertFalse(lines.mapped('full_reconcile_id'))
+
+    def test_02_import_adyen_credit_fees(self):
+        """ Import an Adyen statement with credit fees """
         self._test_statement_import(
             'account_bank_statement_import_adyen',
             'adyen_test_credit_fees.xlsx',
             'YOURCOMPANY_ACCOUNT 2016/8', self.journal)
```
Comment (MPishpecki) [inline account_bank_statement_import_adyen/tests/test_import_adyen.py]: please remove self.journal

File: account_bank_statement_import_adyen/tests/test_import_adyen.py (hunk 1 above)
Comment (StefanRijnhart) [reply account_bank_statement_import_adyen/tests/test_import_adyen.py]: :+1: can you commit the fix yourself?

**Comment by StefanRijnhart** [comment]
Thanks for the improvements! Can you squash that commit?

**Comment by MPishpecki** [comment]
> Thanks for the improvements! Can you squash that commit?

Done

File: account_bank_statement_import_adyen/models/account_bank_statement_import.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                        _('Selected journal Merchant Account does not match '
+                          'the import file Merchant Account '
+                          'column: %s') % account_number)
+                self = self.with_context(journal_id=journal.id)
+        return super(AccountBankStatementImport, self)._find_additional_data(
+            currency_code, account_number)
+
+    @api.model
+    def balance(self, row):
+        return -(row[15] or 0) + sum(
```
Comment (RodrigoBM) [inline account_bank_statement_import_adyen/models/account_bank_statement_import.py:42]: If row position 15,16, 17, 18, 19, 20 is necessary add float conversion

return -(float(row[15]) if row[15] else 0.0) + sum(
            float(row[i]) if row[i] else 0.0
            for i in (16, 17, 18, 19, 20))

File: account_bank_statement_import_adyen/models/account_bank_statement_import.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            row[i] if row[i] else 0.0
+            for i in (16, 17, 18, 19, 20))
+
+    @api.model
+    def import_adyen_transaction(self, statement, statement_id, row):
+        transaction_id = str(len(statement['transactions'])).zfill(4)
+        transaction = dict(
+            unique_import_id=statement_id + transaction_id,
+            date=fields.Date.from_string(row[6]),
```
Comment (RodrigoBM) [inline account_bank_statement_import_adyen/models/account_bank_statement_import.py:68]: row[6] is datetime is necessary row[6].strftime('%Y-%m-%d')

File: account_bank_statement_import_adyen/models/account_bank_statement_import.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                if not statement:
+                    statement = {'transactions': []}
+                    statements.append(statement)
+                    statement_id = '%s %s/%s' % (
+                        row[2], row[6].strftime('%Y'), int(row[23]))
+                    currency_code = row[14]
+                    merchant_id = row[2]
+                    statement['name'] = '%s %s/%s' % (
+                        row[2], row[6].year, row[23])
+                date = fields.Date.from_string(row[6])
```
Comment (RodrigoBM) [inline account_bank_statement_import_adyen/models/account_bank_statement_import.py:118]: row[6] is datetime is necessary row[6].strftime('%Y-%m-%d')

File: account_bank_statement_import_adyen/models/account_bank_statement_import.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                if not statement.get('date') or statement.get('date') > date:
+                    statement['date'] = date
+
+                row[8] = row[8].strip()
+                if row[8] == 'MerchantPayout':
+                    payout -= self.balance(row)
+                else:
+                    balance += self.balance(row)
+                self.import_adyen_transaction(statement, statement_id, row)
+                fees += sum(
```
Comment (RodrigoBM) [inline account_bank_statement_import_adyen/models/account_bank_statement_import.py:128]: If row position 17, 18, 19, 20 is necessary add float conversion
```python
fees += sum(
                    float(row[i]) if row[i] else 0.0
                    for i in (17, 18, 19, 20))
```

File: account_bank_statement_import_adyen/models/account_bank_statement_import.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+            raise ValueError(
+                'Not an Adyen statement. Did not encounter header row.')
+
+        if fees:
+            transaction_id = str(len(statement['transactions'])).zfill(4)
+            transaction = dict(
+                unique_import_id=statement_id + transaction_id,
+                date=max(t['date'] for t in statement['transactions']),
+                amount=-fees,
+                name='Commission, markup etc. batch %s' % (int(row[23])),
```
Comment (RodrigoBM) [inline account_bank_statement_import_adyen/models/account_bank_statement_import.py:142]: row[23] not necessary conversion to int, is a problem if row[23] is a text

`name='Commission, markup etc. batch %s' % (row[23])`

**Review by RodrigoBM — COMMENTED** [review]
Hi @MPishpecki this problems in my testing

File: account_bank_statement_import_adyen/readme/DESCRIPTION.rst [hunk 9]
```diff
@@ -0,0 +1,10 @@
+======================
+Adyen statement import
+======================
+
+This module processes Adyen transaction statements in xlsx format. You can
```
Comment (NL66278) [inline account_bank_statement_import_adyen/readme/DESCRIPTION.rst:1]: Please document what file exactly is handled by this module. I am trying to convert this to 13.0, so to test downloaded the payment accounting report (after login to Ayden: Finance ==> Downloadable reports ==> Payment Accounting). That has 34 columns. Your test file has 31 columns, as is also tested in the code. I cannot find a report on the Ayden site that looks like the one your code is expecting.

**Comment by NL66278** [comment]
Question: why use the xlsx report, leading to extra requirements, instead of the csv report?

**Comment by NL66278** [comment]
And please remove the empty files in the readme directories.

File: account_bank_statement_import_adyen/readme/DESCRIPTION.rst (hunk 9 above)
Comment (NL66278) [reply account_bank_statement_import_adyen/readme/DESCRIPTION.rst:1]: OK I found it. It is the settlement details report from the merchant level. Please document this in the readme.

**Comment by hbrunn** [comment]
runboat build is fixed in https://github.com/OCA/bank-statement-import/pull/468

**Comment by NL66278** [comment]
@hbrunn Hi Holger, could you rebase this branch to solve the conflicts?

**Comment by hbrunn** [comment]
I can't really solve conflicts here because this branch is part of a very tightly pinned buildout. Did the rebase on a new branch and closed this PR

ping @MPishpecki

**Pull request closed without merging**